flask_app/blueprints/api/warnings.py

```diff
@@ -22,7 +22,9 @@ def add_warning(message:str, filename:str=None, lineno:int=None, test_id:int=Non
 
     warning = Warning.query.filter_by(session_id=session_id, test_id=test_id, lineno=lineno, filename=filename, message=message).first()
     if warning is None:
-        if obj.num_warnings < current_app.config['MAX_WARNINGS_PER_ENTITY']:
+        num_distinct_warnings = db.session.query(func.Count(distinct(Warning.lineno and Warning.filename and Warning.message)))\
```

Comment:
`Warning.lineno and Warning.filename and Warning.message` eventually ends up as either `True` / `False` or as the last of the expressions given... you want to use comas instead...

Reply:
Better: `db.session.query(Warning.lineno, Warning.filename, Warning.message).filter_by(...).distinct().count()